libmisc/limits.c

```diff
@@ -68,7 +68,7 @@ static int setrlimit_value (unsigned int resource,
 			return 0;
 		}
 		longlimit *= multiplier;
-		limit = (rlim_t)longlimit;
+		limit = longlimit;
```

Comment:
```c
$ grepc -tfd setrlimit_value
./libmisc/limits.c:43:
static int setrlimit_value (unsigned int resource,
                            const char *value,
                            unsigned int multiplier)
{
	...
	rlim_t limit;

	...

		long longlimit = strtol (value, &endptr, 10);
		...
		longlimit *= multiplier;
		limit = longlimit;
		if (longlimit != limit)
		{
			/* FIXME: Again, silent error handling...
			 * Wouldn't screaming make more sense?
			 */
			return 0;
		}
	}

	...
	return 0;
}
```

It transforms the value from a `long long` to `rlim_t`, which together with the next line ensures that the value isn't changed after converting to `rlim_t`.  The cast did nothing other than clarify the operation, at the cost of being less safe (since casts disable most compiler warnings).

Reply:
But wouldn't that comparison also entail some type of conversion? And if the first conversion did something wrong, why wouldn't the second one do the same?

Reply:
For simplicity, I'll use small numbers, even if long is huge.  Let's imagine that `long longlimit` is 16 bits and `rlim_t limit` is 8 bits.

If `longlimit` has the value 256, the line `limit = longlimit` will store 0 in `limit`.  Then in the next line where we compare both, it will execute this: `if (256!= 0)` (and both will be converted to the larger type, so the values will be kept), which is true.

Reply:
Thanks for the explanation.